tests/samplers_tests/test_brute_force.py

```diff
@@ -127,6 +127,37 @@ def objective(trial: Trial) -> float:
         assert a in expected_suggested_values
 
 
+def test_study_optimize_with_pruned_trials() -> None:
+    def objective(trial: Trial) -> float:
+        a = trial.suggest_int("a", 0, 2)
+
+        if a == 0:
+            trial.suggest_float("b", -1.0, 1.0, step=0.5)
+        elif a == 1:
+            trial.suggest_categorical("c", ["x", "y", None])
+
+        raise optuna.TrialPruned
```

Comment:
:nit: Wouldn't we need a `return` statement?

Reply:
I think it's okay because the pruned state is set in the `catch TrialState.Pruned` line of the [`_optimize` function.](https://github.com/optuna/optuna/blob/master/optuna/study/_optimize.py#L199-L204).

Reply:
I think this objective function will always be `Pruned`, so it would be better to make the pruning stochastic.

Reply:
I have updated the test to include `COMPLETE` trials. PTAL.

Reply:
LGTM! 
@gen740 Could you take another look? If you approve, I will merge this PR.